### PR TITLE
Possible fix for physics crashes

### PR DESCRIPTION
**Pull request by NolaDonato**
This patch synchronizes access to the native physics step() function and does not allow it to be called if it is still processing the last frame. This MAY fix some of the crashes in the physics apps. It needs stress testing before we can declare it a true fix.

GearVRF DCO signed off by: Nola Donato nola.donato@samsung.com

Fixes for https://github.com/Samsung/GearVRf/issues/1100

### CODE REVIEW

File: GVRf/Framework/framework/src/main/jni/objects/components/render_data.cpp [hunk 1]
```diff
@@ -36,9 +36,14 @@ const RenderPass* RenderData::pass(int pass) const {
...
-    mesh_ = mesh;
-    mesh->add_dirty_flag(dirty_flag_);
-    *dirty_flag_ = true;
+    if (mesh != nullptr) {
+        mesh_ = mesh;
+        mesh->add_dirty_flag(dirty_flag_);
+        *dirty_flag_ = true;
+    }
+    else {
+        LOGE("RenderData::set_mesh mesh cannot be NULL");
```
Comment (liaxim) [inline GVRf/Framework/framework/src/main/jni/objects/components/render_data.cpp]: It should probably std::terminate in addition to the log trace. Or just use the FAIL macro from gvr_log.h.

File: GVRf/Extensions/gvrf-physics/src/main/java/org/gearvrf/physics/GVRWorld.java [hunk 2]
```diff
@@ -49,6 +51,8 @@ public GVRWorld(GVRContext gvrContext) {
     public GVRWorld(GVRContext gvrContext, GVRCollisionMatrix collisionMatrix) {
         super(gvrContext, NativePhysics3DWorld.ctor());
         mHasFrameCallback = false;
+        mIsProcessing = false;
```
Comment (liaxim) [inline GVRf/Extensions/gvrf-physics/src/main/java/org/gearvrf/physics/GVRWorld.java]: What is setting mIsProcessing to true? Are there changes missing from this pull request?

File: GVRf/Framework/framework/src/main/jni/objects/components/render_data.cpp [hunk 3]
```diff
@@ -36,9 +36,14 @@ const RenderPass* RenderData::pass(int pass) const {
 }
 
 void RenderData::set_mesh(Mesh* mesh) {
-    mesh_ = mesh;
-    mesh->add_dirty_flag(dirty_flag_);
-    *dirty_flag_ = true;
+    if (mesh != nullptr) {
```
Comment (liaxim) [inline GVRf/Framework/framework/src/main/jni/objects/components/render_data.cpp]: This nullptr check prevents in the crash in NativePhysics3DWorld.step. Though I believe the better fix is to synchronize all access to GVRWorld.mRigidBodies.

**Comment by liaxim** [comment]
This is much more than a crash fix by now and is not affecting just the physics extension. Some things could use few lines of explanation what problems are being fixed and how are those reproduced/verified as being fixed.

Since end users use the Java APIs to deal with Bullet there is still some locking needed on the Java side too and there will be overlap with the locking on the native side. This may very well fix all the problems with gvr-simplephysics but requires extensive testing better done after the release. I already have few comments and will have some more in the future.

**Comment by liaxim** [comment]
gvr-simplephysics with the modifications to cause more crashes is not crashing with these changes in.

**Comment by NolaDonato** [comment]
This pull request will accumulate everything necessary to make the demos work. I don't think we should check this in now. There are still more bugs in physics. This patch just fixes obviously bogus stuff. Neither physics nor GearVRf locks around updating transforms but they can both update them simultaneously. This will never be thread safe and we need to fix it. I  think we should hold the release to fix physics.

**Comment by liaxim** [comment]
If we keep expanding the scope of the release then more things will crop up in the meantime and we are not going to have a release any time soon.

**Comment by NolaDonato** [comment]
replaced by #1141 and #1140

**Pull request closed without merging**